### ISSUE TITLE
Videos randomly disappearing

### ISSUE THREAD

**Issue by rem0g** [Signbank/Global-signbank#1453]
Some glosses has disappeared videos for example: 

https://signbank.cls.ru.nl/dictionary/gloss/47361

https://signbank.cls.ru.nl/dictionary/gloss/47475

Some glosses has wrong still image from the video. 

This is happening everywhere, can you take a look into that?

**Comment by rem0g** [comment]
At Glos AFBROKKELEN (https://signbank.cls.ru.nl//dictionary/gloss/46883), I am seeing this: 


      {
                    "ID": "211108",
                    "Index": "0",
                    "Description: Dutch": "afbrokkelen",
                    "Description: English": "crumble",
                    "Link": "https://signbank.cls.ru.nl//dictionary/protected_media/glossvideo/NGT/AF/AFBROKKELEN-46883.mp4"
                },
                {
                    "ID": "208048",
                    "Index": "1",
                    "Description: Dutch": "afbrokkelen",
                    "Description: English": "crumble",
                    "Link": "https://signbank.cls.ru.nl//dictionary/protected_media/glossvideo/NGT/AF/AFBROKKELEN-46883.mp4"
                }
                
                
                Both has same link and that should not be the case.

**Comment by susanodd** [comment]
@rem0g the university is closed this week and next. We won't be able to work on this until then.
We are all on vacation now.

**Comment by susanodd** [comment]
Can you increase the amount if time between updating/uploading and retrieving?

Like maybe increase it to 10 minutes? There was previously a problem that the time between operations was too frequent.
With videos in could be that a previous operation was not completed before a new one is performed. (It makes a difference if the file system is continuously writing new files. The transactions to create objects are faster than the file system.

**Comment by susanodd** [comment]
I had this problem locally on my own computer using iCloud for storage.
But I could not repeat it on the real server.
It was that the files had not been completely copied to iCloud. So they "disappeared" like that. They were actually on disc, but had names with "." before the filename, so not visible. This was on Apple's Unix.

**Comment by rem0g** [comment]
Yes I could do that but i would rather for Signbank side do something about the transactions, for example for incoming transactions I always enqueue them in a list instead of executing them immediately. When the server has finished processing a certain transaction then it could handle the next one. 

Would that be something you can work on?

**Comment by susanodd** [comment]
@rem0g that sounds like an interesting approach. I will discuss that with @vanlummelhuizen how to implement that. He is the Django expert. A queuing mechanism.

**Comment by susanodd** [comment]
> Another example:
> 
> https://signbank.cls.ru.nl//dictionary/gloss/47336
> 
> When glos video has been deleted, then NME video has moved from NME to glos.

@rem0g How are you deleting the video?

(There are some "signals" when objects are deleted. These may or may not move or delete video files. It would help in debugging to know what commands have been done.)
(When a gloss is deleted, the video files are not deleted.)

I implemented a lot of code in November/December for managing the video files. There are pull requests for these. But nobody has reviewed them yet. The intention is that the dataset manager can inspect what is in the file system. That also allows to retrieve the videos from deleted glosses. The gloss IDs are not reused, so new videos should not have any interference with deleted glosses, since they always have the ID in the filename and these are not reused.)

**Comment by susanodd** [comment]
@rem0g for this gloss:

https://signbank.cls.ru.nl/dictionary/gloss/47361

the NME video is not in the correct format!

(On Firefox, it shows that it is not supported.)

Recall that you asked us to not test for MP4 anymore. Thus it can be that incorrect formats are causing problems.
(The images cannot be generated for videos in the wrong format.)

**Comment by susanodd** [comment]
@rem0g here are the gloss video objects for AFBROKKELEN, yes, you can see that the same filename appears multiple times, for different GlossVideo objects with various perspectives and NME set.

![afbrokkelen-46883-glossvideo-objects](https://github.com/user-attachments/assets/6318c41a-7bd8-405f-9756-509ad7b96735)

**Comment by susanodd** [comment]
Is it possible something was wrong with the permissions on your source file? Or that it was a symbolic link?
It looks like something was wrong with the source file that it kept being bound to different objects.
Also, all of the files attached to the objects have very similar time stamps.

**Comment by susanodd** [comment]
@vanlummelhuizen can you help on this issue?

**Comment by susanodd** [comment]
@rem0g there are hundreds of videos with the wrong filename as you point out.

Did you change anything in your script?
I copied the newest database to signbank-test in order to inspect the filenames (in the objects).
The last time I checked filenames (end of November) everything was as expected.
It may have started when you can upload three videos at the same time in the API?

**Comment by susanodd** [comment]
I need help with this issue.

One of the glosses that is messed up is STRUISVOGEL


![STRUISVOGEL-perspective-videos-admin](https://github.com/user-attachments/assets/adefe711-386e-4d5d-b67b-e8b0215052e3)

```
stat ../writable/glossvideo/NGT/ST/STRUISVOGEL-45874.mp4
  File: ../writable/glossvideo/NGT/ST/STRUISVOGEL-45874.mp4
  Size: 1105797   	Blocks: 2158       IO Block: 131072 regular file
Device: 20007dh/2097277d	Inode: 864571      Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/  ubuntu)   Gid: ( 1001/wwwsignbank)
Access: 2025-01-06 08:10:59.506318499 +0000
Modify: 2024-12-01 05:01:46.404138548 +0000
Change: 2024-12-20 13:58:20.693990022 +0000
 Birth: 2024-12-01 05:01:46.386138096 +0000
```

**Comment by susanodd** [comment]
I don't know how to solve this.

Hopefully, @vanlummelhuizen can have a go.

**Comment by rem0g** [comment]
Hello, thank you for looking into the issue. For me the issue is not clear yet but now I know it's not caused by my script as everything relies on Signbank API. For uploading videos i do use this API: 

/dictionary/api_update_gloss/{glossid}/video

That's it. 

As for NME video upload i do use: 

/dictionary/api_create_gloss_nmevideo/{datasetid}/{glossid}/}

And for deleting NME video: 

/dictionary/api_delete_gloss_nmevideo/{datasetid}/{glossid}/{videoid}/

For every time if i want to upload a certain NME video i do execute api_delete_glos_nmevideo first, but for that i obtain unique ID from the nme video and then delete that and then upload the new NME video.

**Comment by uklomp** [comment]
Hi all, I added 'blocking' to indicate extra extra priority. If there's something on our end we can do, please let us know.

**Comment by susanodd** [comment]
I'm not able to solve this myself. I am aware of this problem for a long time, but it was on a local server running on iCloud. So I chalked that up to Apple quirks.

There are quite a few messed up video objects/ files now.

**Comment by susanodd** [comment]
@rem0g can you stop deleting the videos on your side? Because objects are referring to the same file, this is causing a shared file to be deleted.

I still think this is due to the API commands happening too fast for the file system. But once a file ends up shared by different objects, that escalates the problem, domino effect.

**Comment by susanodd** [comment]
@rem0g I put the database (per yesterday) onto `signbank-susan `where I have a pull request branch running with extra commands for the admin. (For filtering on NME, Perspective, and wrong filename).

Out of curiosity, I looked at the gloss video history. Here you can see the names of the files that were uploaded. The source file and the (desired) target file.

It looks like your naming convention with L, M, R in the filenames, that you are also using M for videos that are NME videos.

![glossvideohistory_recent](https://github.com/user-attachments/assets/ee2c0573-f9f0-4b25-86cb-e8188b482640)

**Comment by susanodd** [comment]
I have a copy of the database on signbank-susan. (No videos, but you can browse the video objects / filenames / paths) in the admin.

I am working on filters to enhance the admin in order to detect and fix video problems. #1398 

I added filters for NME and Perspective, plus a filter for "wrong filename".

Now it's possible to query on those wrong filenames.

For NME videos with the wrong filename, there are 1366 results!!!!!

**Comment by vanlummelhuizen** [comment]
I have looked into this problem, but without succes so far. 

My hypothesis is that, because GlossVideoNME is a subclass of GlossVideo, the delete function of GlossVideoNME is executed as if it was a GlossVideo.

<details>
<summary>Things I found out about FRIKANDEL-A (gloss id 47336)</summary>

The contents of the corresponding video file were modified on December 18 (the file's birth), while it's metadata were changed on December 20 (change of file name, move to different location, etc):

```
root@signbank-new:/var/www/repo# stat ../writable/glossvideo/NGT/FR/FRIKANDEL-A-47336.mp4
  File: ../writable/glossvideo/NGT/FR/FRIKANDEL-A-47336.mp4
  Size: 1435252   	Blocks: 2816       IO Block: 131072 regular file
Device: 20007dh/2097277d	Inode: 623811      Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/  ubuntu)   Gid: ( 1001/wwwsignbank)
Access: 2025-01-08 23:26:03.444677917 +0000
Modify: 2024-12-18 18:18:40.134785862 +0000
Change: 2024-12-20 13:59:33.609808729 +0000
 Birth: 2024-12-18 18:18:40.124785614 +0000
```
---
An NME video was deleted *and* created again on December 18:
```
sqlite> select * from dictionary_glossrevision where field_name like 'nmevideo_%' and gloss_id = 47336;
id|time|field_name|old_value|new_value|gloss_id|user_id
154385|2024-11-14 11:42:07.510864|nmevideo_create||FRIKANDEL-A-47336_nme_0.mp4|47336|323
155427|2024-11-14 12:40:33.165356|nmevideo_create||FRIKANDEL-A-47336_nme_1.mp4|47336|323
159004|2024-11-14 15:57:19.854011|nmevideo_delete|FRIKANDEL-A-47336_nme_0.mp4||47336|323
159005|2024-11-14 15:57:20.152481|nmevideo_create||FRIKANDEL-A-47336_nme_0.mp4|47336|323
159008|2024-11-14 15:57:20.377890|nmevideo_delete|FRIKANDEL-A-47336_nme_1.mp4||47336|323
160029|2024-11-14 16:41:34.021974|nmevideo_delete|FRIKANDEL-A-47336_nme_0.mp4||47336|323
160030|2024-11-14 16:41:34.234272|nmevideo_create||FRIKANDEL-A-47336_nme_0.mp4|47336|323
282341|2024-12-18 18:18:38.888102|nmevideo_delete|FRIKANDEL-A-47336_nme_0.mp4||47336|323
282342|2024-12-18 18:18:39.181953|nmevideo_create||FRIKANDEL-A-47336_nme_0.mp4|47336|323
282345|2024-12-18 18:18:39.881494|nmevideo_delete|FRIKANDEL-A-47336_nme_0.mp4||47336|323
282346|2024-12-18 18:18:40.147739|nmevideo_create||FRIKANDEL-A-47336_nme_0.mp4|47336|323

```
and in the logs
```
[pid: 337|app: 0|req: 358/3269] 10.208.155.251 () {60 vars in 1148 bytes} [Wed Dec 18 19:18:38 2024] GET /dictionary/api_delete_gloss_nmevideo/5/47336/26403/ => generated 61 bytes in 21 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 3)
[pid: 337|app: 0|req: 359/3270] 10.208.155.251 () {64 vars in 1234 bytes} [Wed Dec 18 19:18:39 2024] POST /dictionary/api_create_gloss_nmevideo/5/47336/ => generated 82 bytes in 56 msecs (HTTP/1.1 200) 5 headers in 147 bytes (88 switches on core 1)
[pid: 350|app: 0|req: 1469/3271] 10.208.155.251 () {60 vars in 1148 bytes} [Wed Dec 18 19:18:39 2024] GET /dictionary/api_delete_gloss_nmevideo/5/47336/26403/ => generated 98 bytes in 14 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 0)
(...)
[pid: 337|app: 0|req: 361/3274] 10.208.155.251 () {60 vars in 1151 bytes} [Wed Dec 18 19:18:39 2024] GET /dictionary/api_delete_gloss_nmevideo/5/47336/211214/ => generated 61 bytes in 20 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 3)
[pid: 350|app: 0|req: 1471/3275] 10.208.155.251 () {64 vars in 1234 bytes} [Wed Dec 18 19:18:40 2024] POST /dictionary/api_create_gloss_nmevideo/5/47336/ => generated 82 bytes in 44 msecs (HTTP/1.1 200) 5 headers in 147 bytes (55 switches on core 2)
[pid: 341|app: 0|req: 557/3276] 10.208.155.251 () {60 vars in 1151 bytes} [Wed Dec 18 19:18:40 2024] GET /dictionary/api_delete_gloss_nmevideo/5/47336/211214/ => generated 98 bytes in 13 msecs (HTTP/1.1 200) 5 headers in 147 bytes (1 switches on core 3)
```

</details>

**Comment by rem0g** [comment]
NME videos are indeed deleted via API url: 

    url = f"https://signbank.cls.ru.nl/dictionary/api_delete_gloss_nmevideo/5/{gloss_id}/{videoid}/"


the videoid is obtained via: 
    url = f"https://signbank.cls.ru.nl/dictionary/get_gloss_data/5/{gloss_id}"
    
    and then 
    
        nme_videos = gloss_info.get("NME Videos", [])

then we execute deletion via: 

    url = f"https://signbank.cls.ru.nl/dictionary/api_delete_gloss_nmevideo/5/{gloss_id}/{videoid}/"


When we want to replace NME video, we call that api_delete_gloss_nmevideo and then upload new nme_video again via: 

    url = f"https://signbank.cls.ru.nl/dictionary/api_create_gloss_nmevideo/5/{gloss_id}/"
  # Endpoint of intermediary server

**Comment by vanlummelhuizen** [comment]
@rem0g In the procedure you describe above you seem to delete a NME video twice via https://signbank.cls.ru.nl/dictionary/api_delete_gloss_nmevideo/5/{gloss_id}/{videoid}/. Is that the case? I would think only one call should be enough.

**Comment by rem0g** [comment]
It’s executed once just before new nme video creation. The first one is
just a “title” of the workflow so it’s done one time though.

**Comment by Woseseltops** [comment]
I'm not sure if we're all on the same page regarding this problem. Let me summarize:
* For a gloss like FRIKANDEL-A, we would expect several video files, like the main `FRIKANDEL-A-47336.mp4`, but also perspective variants like `FRIKANDEL-A-47336_left.mp4`, and NME videos like `FRIKANDEL-A-47336_nme_0.mp4`
* Signbank thinks both perspective videos and NME videos exist for FRIKANDEL_A... but when you request the file paths of these videos, it just returns the main video path `FRIKANDEL-A-47336.mp4` . 
* On top of this, this main video file `FRIKANDEL-A-47336.mp4` is NOT the original main video, but an NME video.
* Looking at the video files at the server, we see a lot of backups (not sure of what exactly), but no `FRIKANDEL-A-47336_left.mp4` or `FRIKANDEL-A-47336_nme_0.mp4` .

Unlikely as it is, this seems to have happened:
* An NME video has replaced the original main video
* All video paths, for both perspective and NME videos, have been changed to link to this main video 
* All additional videos that once existed, but now were not linked to anything anymore, have been deleted

**Comment by susanodd** [comment]
Pointed out earlier, in the video history (admin), some video files that are uploaded are using the M... for the NME video as well, so name of a source video is the same for the normal video as an NME video. For videos larger than 2.5 mb (Django quote above) these are in temp storage. It could happen that somehow the uploaded file in temp storage is being used by multiple objects because it has the same name.

**Comment by Woseseltops** [comment]
Two new discoveries:
* I just tried to reproduce the problem by adding and removing NME videos video the API a few times, but no luck... it all worked as intended :/ . Gomer must have done something differently, but I cannot imagine what.
* The other example glosses given by Gomer, #BMW and #PSV, don't show perspective and NME videos, but that's because of an unrelated problem (the hashtag in the name). The actual videos linked show the same pattern as FRIKANDEL-A

**Comment by Woseseltops** [comment]
> Pointed out earlier, in the video history (admin), some video files that are uploaded are using the M... for the NME video as well, so name of a source video is the same for the normal video as an NME video. For videos larger than 2.5 mb (Django quote above) these are in temp storage. It could happen that somehow the uploaded file in temp storage is being used by multiple objects because it has the same name.

I don't think this is what is happening here: mixing up files from temporary storage would lead to duplicates, all with correct file paths. Instead, we see that the URL paths stored in the database themselves are incorrect, and that the files at the correct path don't exist.

**Comment by susanodd** [comment]
> > Pointed out earlier, in the video history (admin), some video files that are uploaded are using the M... for the NME video as well, so name of a source video is the same for the normal video as an NME video. For videos larger than 2.5 mb (Django quote above) these are in temp storage. It could happen that somehow the uploaded file in temp storage is being used by multiple objects because it has the same name.
> 
> I don't think this is what is happening here: mixing up files from temporary storage would lead to duplicates, all with correct file paths. Instead, we see that the URL paths stored in the database themselves are incorrect, and that the files at the correct path don't exist.

Could it happen if a transaction doesn't complete? The object creations don't wait on the file system.

Or if the Django implementation (of Django) is somehow using symbolic links?

No idea, I'll stop. It's just weird.

When running it on iCloud, I had the same problem. It turned out the files were there, but with "." before them and had not been uploaded to iCloud yet.

**Comment by rem0g** [comment]
What would happen if you execute the transaction with 10 different glosses
at short timespan?


I am going to run the script on ten glosses soon.

Gomer

**Comment by rem0g** [comment]
I am yet to see weird results but for this API I have encountered a bug:

I have to send POST update to api_update_gloss/{gloss}/video twice as the videos dont appear the first time i POST the videos, after 5 seconds again POST-ing it they appear again. 

I don't know if that's related to this issue.

Gomer

**Comment by rem0g** [comment]
I am just going ahead and re-upload all base glos video files, but with sleep of 5 sec between uploads. If that's finished then i will start with NME video delete and re-uploads.

**Comment by Woseseltops** [comment]
> What would happen if you execute the transaction with 10 different glosses at short timespan?

I expect nothing special, except maybe database locks if it's faster than the database can handle.

> I have to send POST update to api_update_gloss/{gloss}/video twice as the videos dont appear the first time i POST the videos, after 5 seconds again POST-ing it they appear again.

I think unrelated. Respond to this comment with your fav emoji if you want me to create a separate issue for it :)

> I am just going ahead and re-upload all base glos video files, but with sleep of 5 sec between uploads. If that's finished then i will start with NME video delete and re-uploads.

Apologies that this is needed, I realize the problem is on our side :(.

I think our best chance of fixing the problem in the current issue is being able to reproduce it. If you notice the disappearing/overwriting of videos again, would be great if you could (for one example gloss) explain exactly which commands you run so that I can try to make the behavior appear at will.

**Comment by rem0g** [comment]
Ok thank you! We will see what happens and no need to have seperate issue for that, for now :)

**Comment by rem0g** [comment]
Still reuploading as I also have script written to double check if video is there at gloss, if not then reupload again. 

I came across this for example: 

https://signbank.cls.ru.nl//dictionary/gloss/45834

Gloss video is in NME index, but it should not be there. 

I have theory what is happening: 

Gloss Video URL is https://signbank.cls.ru.nl/dictionary/protected_media/glossvideo/NGT/RA/RAT-A-45834.mp4?v=20250111170412

NME video URL is also:                     "Link": "https://signbank.cls.ru.nl//dictionary/protected_media/glossvideo/NGT/RA/RAT-A-45834.mp4"


When i run API to delete the NME video index 0, it will remove the gloss video. 

I think that looks a reasonable cause.

My proposal for a bugfix: 

Disable api_delete_gloss_nmevideo as we can simply overwrite existing NME video at index 0 for example. 

API documentation states this: 

Updates the details of a 'non-manual element' video for a gloss. To replace the video, delete it using api_delete_gloss_nmevideo and create a new NME video using api_create_gloss_nmevideo.

So that requirement will not be need anymore. 

For now I have disabled NME video upload script, when the bug if recognized and fixed then I will re-activate it.

**Comment by susanodd** [comment]
Yes, it is true that it overwrites the video with the same index number.
Originally, the indexing started at 1. I modified that to also allow index 0 since you were using that in your API and creating them with 0.
Originally, before the API, it was starting them at 1 because the template forms do this with number fields.

When you used to use the "zip" upload, the video files were overwritten. The backup system was avoided altogether.
The problems kind of started with the backup system came into use.

The normal videos do not delete the video files. But the subclasses do delete them.
Since somehow (we don't know how) the NME and perspective videos were being given the same file name inside the object as the normal video, then when you delete these, it was deleting the normal video file.

If you just stop using delete, it should just work, since it overwrites them. Did you try this?
Can you start numbering with index 1?
It could be something that the "0" index is somehow ending up matching "Null" and the code thinks the object is a normal video.
(If coercion on values is doing something.)

**Comment by susanodd** [comment]
Another try, can you make the index start at 1 of the NME video?

Using 0 can be dangerous, since it is also sometimes the same as False or None in the database. Because things are strings in the API it could be a problem with type coercion. Also, Django automatically converts objects to their "id" when they are passed around. They only remain an actual object in the template if it's via context variables.

**Comment by susanodd** [comment]
> Using 0 as replacement for gloss sounds weird as I understand the index number is for NME videos only.

The only difference between a normal GlossVideo object and the (subclass) GlossVideoNME object is that the NME object has an "offset" field added to it. Otherwise they have all the same fields in the objects. In the model, the default offset is 1.

I was merely wondering out loud if it's possible that an offset with value 0 is somehow "empty" in the database. Thereby making Django think that an NME object is a normal video object. (I have no idea if that could happen. I was wondering out loud. That's why I suggested to just use 1. Then it's clear it would not be empty.)

> Nothing has been sent to Signbank via API after last reupload of all videos before my vacation (happened past friday or so). NME autoupload and api_delete_gloss_nmevideo API call has been disabled since then too.
> 
> Per @uklomp comment everything was fine this morning, so something has happened on Signbank server outside of our scope? I hope @Woseseltops can solve this issue this week as we are nearing the end of our Signbank project.

We haven't changed anything on our side. Only the CNCZ could have changed something (the university owns the websites. I have no idea whether there are any limits to how much data we can store.)

**Comment by Woseseltops** [comment]
This issue has been eating up nearly all of my Signbank time (and more), but I can't figure it out; the code looks okay as far as I can see. My main approach therefore is to try to reproduce the problem, using these API endpoints in various orders:

* `/dictionary/api_update_gloss/{glossid}/video` (center video, but also perspective videos)
* `/dictionary/api_create_gloss_nmevideo/{datasetid}/{glossid}/`
* `/dictionary/api_delete_gloss_nmevideo/{datasetid}/{glossid}/{videoid}/`

Any other endpoints you are using, @rem0g ?

**Comment by Woseseltops** [comment]
After a longer session with @vanlummelhuizen , we came to the conclusion that the following is likely happening:

* Some mysterious process, let's call it 'process X', wants to rename or move the **main** video
* To do this, it incorrectly requests ALL videos with `GlossVideo.objects.filter(gloss=glossid)`
* It renames each video with the new, same path, and stores this in the database. Every time this happens, the new file overwrites the previous one, so only the last one remains.

The question now is, where in the code does process X live and what does it do? The video path consists of the dataset, the lemma and the private key, so it must be related to changing any of these.

**Comment by susanodd** [comment]
> The question now is, where in the code does process X live and what does it do? The video path consists of the sign language, dataset, the lemma and the private key, so it must be related to changing any of these.

Both of the signal methods for updating the dataset or lemma of a gloss move the video files! As well as updating the annotations.

The dataset signal processor will move all the videos of the dataset.

**Comment by uklomp** [comment]
Hi all, I don't mean to put any extra pressure on you because I believe you're on top of it, but is it possible to have an indication when this is solved? We can't proceed with the last phase in the project right now, so it would be good to know if you're thinking this will take another week or another month  -or that you have no idea. Thanks!

**Comment by susanodd** [comment]
We have not been able to duplicate the error when we do it ourselves.


If you go to the Dataset Manager, Manage Video Storage, you can see all the paths of all the video objects for the dataset (NGT is dataset 5)

https://signbank.cls.ru.nl/datasets/checks/5

The page takes a while to load.
The first panel shows "video file paths" that do not have a gloss pointing to them. (Those are videos of deleted glosses, the files remain in the file system.) (I was hoping to find SINTERKLAAS here.)

Further down the page, you can see the file paths for i.e., the perspective, NME, and normal videos. (Then you can see to what extent the paths are messed up.)

**Comment by vanlummelhuizen** [comment]
> Hi all, I don't mean to put any extra pressure on you because I believe you're on top of it, but is it possible to have an indication when this is solved? We can't proceed with the last phase in the project right now, so it would be good to know if you're thinking this will take another week or another month -or that you have no idea. Thanks!

@uklomp We are working on it, but as @susanodd mentions, we have not been able to replicate the problem. Nor can we pinpoint it in the logs or code base. This makes it extremely difficult to give a prognosis. 

I will discuss this (and possible workarounds) with @Woseseltops tomorrow. We will keep you posted.

**Comment by vanlummelhuizen** [comment]
Perhaps I found the code that causes this. Question for @uklomp and @rem0g : do you remember that one of you changed the default language of the NGT dataset on Tuesday January 14 around 13:15? Around that time I see a POST request to `/datasets/change_details/5`, where 5 is the ID of the NGT dataset.

I will explain what I found below.

<details>

A POST to `/datasets/change_details/5` calls https://github.com/Signbank/Global-signbank/blob/759d126d7971afa27b4d51d0a036ff6a1cbdc135/signbank/dictionary/update.py#L2943-L2943 where the dataset is saved with `dataset.save()`. 

This triggers a post_save signal that in turn calls the following function:
https://github.com/Signbank/Global-signbank/blob/759d126d7971afa27b4d51d0a036ff6a1cbdc135/signbank/video/models.py#L1049-L1090

In the last if-block in the code above (line 1083 and further), the `move_video` method of the GlossVideo class (!) is called on every GlossVideo object, including GlossVideoNME and GlossVideoPerspective objects. Note that this is the `move_video`  method in the class GlossVideo and **not** the methods with the same name in either GlossVideoNME or GlossVideoPerspective: https://github.com/Signbank/Global-signbank/blob/759d126d7971afa27b4d51d0a036ff6a1cbdc135/signbank/video/models.py#L832-L871

In the code above, there is a call to `get_video_file_path` (line 838), but since it is the method in GlossVideo, there is no `nmevideo` or `perspective` argument in the call. This results in new_path without any NME of Perspective indicator, making it the same as the neutral video name **and** different from the original. This triggers a file rename, and when the file on disk is renamed, the neutral video file is overwritten by another file. Finally, if this is done is a particular order, e.g. 'left', 'nme', 'right' (alphabetical), you end up with only the right-perspective video that has the file name of the neutral video.

Code of `get_video_file_path`: https://github.com/Signbank/Global-signbank/blob/759d126d7971afa27b4d51d0a036ff6a1cbdc135/signbank/video/models.py#L139-L181

</details>

**Comment by rem0g** [comment]
@Woseseltops asked if we are using only those endpoints, for the script regarding NME video autoupload i am using those endpoints: 

    url = f"https://signbank.cls.ru.nl/dictionary/get_gloss_data/5/{gloss_id}"
    url = f"https://signbank.cls.ru.nl/dictionary/api_delete_gloss_nmevideo/5/{gloss_id}/{videoid}/"
    url = f"https://signbank.cls.ru.nl/dictionary/api_create_gloss_nmevideo/5/{gloss_id}/"


For the gloss video autoupload i am using only: 
    url = f"https://signbank.cls.ru.nl/dictionary/api_update_gloss/{gloss}/video"


@vanlummelhuizen mentioned about /datasets/change_details/5, I havent implemented the endpoint so far in the scripts. To be sure I havent missed anything i have done command 'cat * | grep change_details' on all scripts, but nothing found. It's also not described in API documentation.

I don't really see anything useful with the endpoint change_details as we use endpoints for gloss video and gloss phonology update/retrieval/deletion only, so probably it would be the best if the change_details endpoint would be ..removed?  Either way, if the fix is implemented then i will reupload all base gloss videos again.

**Comment by vanlummelhuizen** [comment]
@rem0g Perhaps the change was done through the front-end.

**Comment by susanodd** [comment]
> [@rem0g](https://github.com/rem0g) Perhaps the change was done through the front-end.

I think you found it!

Just doing a "save" on a dataset object unleashes that signal method that moves the videos. Even if the dataset name hasn't been changed, the dataset has been updated so the method is called.

**Comment by susanodd** [comment]
@uklomp was doing something with the dataset description. She found html code in it. So that could be what triggered it.

**Comment by vanlummelhuizen** [comment]
Update: I have been able to replicate the problem. Apparently there are two bugs:

1. In the `post_save` receiver the current default language and the one from the Change Dataset form are compared incorrectly, i. e. ID vs name: https://github.com/Signbank/Global-signbank/blob/759d126d7971afa27b4d51d0a036ff6a1cbdc135/signbank/video/models.py#L1083-L1083
2. As I wrote yesterday, the GlossVideoPerspective and GlossVideoNME objects are handled as if they are GlossVideo objects: https://github.com/Signbank/Global-signbank/blob/759d126d7971afa27b4d51d0a036ff6a1cbdc135/signbank/video/models.py#L1085-L1087

**Comment by susanodd** [comment]
Can we please add the "acronym" to the updated_fields for the dataset signals? That would have prevented all of this.

**Comment by uklomp** [comment]
Gomer will run the script for uploading all the videos again tonight!

**Comment by rem0g** [comment]
Script has done running but I noticed a lot of glosses dont have base videos, there is a quite large number that is missing videos even after they were uploaded. I'm going to run the script again.

**Comment by susanodd** [comment]
> Script has done running but I noticed a lot of glosses dont have base videos, there is a quite large number that is missing videos even after they were uploaded. I'm going to run the script again.

Can you build in a sufficient time interval to make sure the file system has completed.

Because there is an extreme number of "backup" videos at the moment, all of the backup video objects are also updated.

(There is a pull request for Admin commands to get rid of them, but it has been ignored by the colleagues.)

**Comment by susanodd** [comment]
This gloss:

https://signbank.cls.ru.nl/dictionary/gloss/47683/

The NME video has the same name as the normal video.
